### ISSUE TITLE
Draft preview - JWT token missing on frontend when domains are different.

### ISSUE THREAD

**Issue by prowsejeremy** [payloadcms/payload#4918]
TLDR;

How (if at all) is the preview authentication process via JWT intended to work across multiple TLDs (top level domains)?

Context...

When implementing draft-preview mode, the JWT token created by payload and sent as a cookie header doesn't make it to the frontend. This only happens when the backend and frontend domains differ, for example:
Payload Backend: https://api.localhost/
Nextjs Frontend: https://localhost/
Everything works as expected when running the draft preview example found here:
https://github.com/payloadcms/payload/tree/main/examples/draft-preview
But when these urls are replaced with the example ones above the token no longer reaches the frontend.

From my understanding, configuring the user auth cookie as such:
  ...
    cookies: {
      sameSite: 'none',
      secure: true,
      domain: process.env.PAYLOAD_URL
    }
  ...

should allow the cookie to be sent across differing domains, however this is not the behaviour I am seeing.

I can get preview to work in this context by omitting the JWT token and solely using the preview secret to authenticate, however this to me seems very insecure and is not what is outlined as intended functionality in the Payload docs.

**Comment by DanRibbens** [comment]
Have you added the requesting domain to the ` ``````cors` so that your API trusts the frontend?

The cookie belongs to the API domain which is what sets it in the login endpoint. The frontend doesn't get access to the cookie and doesn't need it I thought.

**Comment by prowsejeremy** [comment]
Hey @DanRibbens, yes the domains have been added to both the carf and cors arrays, in the same way they are configured in the draft-preview example linked in my original post.

Yes I am aware the cookie belongs to the api domain, however the functionality outlined in the above "draft-preview" has that cookie (the JWT auth token) being passed along to the frontend where it is used to verify the login status of the user attempting to query the previewed collection item, see [this file in the example](https://github.com/payloadcms/payload/blob/main/examples/draft-preview/next-app/app/api/preview/route.ts).

In testing this works as intended when both apps are running on `http://localhost:{PORT}` you can log out the token value on the frontend and it matches the issued token from the payload backend. However as stated this is not the case when these root domains differ.

Cheers.

**Comment by ovalice** [comment]
Are you sure that the cookie is properly set? Since it is a 3rd party cookie, it can be blocked by your browser.

**Comment by prowsejeremy** [comment]
Hey @ovalice Yup I am, I can see in the browser window that a cookie is being issued and assigned to the domain defined in the cookies object outlined above. I have also setup both `api.localhost` and  `localhost` to run with ssl enabled as per the http cookie specifications [outlined here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#sect4). Provided I am understanding this correctly!

Here is an example output of the cookie value I am seeing in browser (note that I have attempted configuring this cookie with both api.localhost and localhost as the domain options):

<img width="1440" alt="Screen Shot 2024-01-25 at 4 16 31 PM" src="https://github.com/payloadcms/payload/assets/4566242/dae7063e-58ca-405b-a961-ae5ba9c5d290">

**Comment by prowsejeremy** [comment]
From what I can tell a number of other CMS solutions don't actually use any form of authentication to enable the viewing of preview data. Looking into both [Directus](https://docs.directus.io/guides/headless-cms/live-preview/nextjs.html) and [Strapi's](https://strapi.io/blog/implementing-previews-with-next-applications-using-a-strapi-backend) approach reveals they mainly rely on a `PREVIEW_SECRET` key to validate a users permissions to view draft content. I got excited when I saw a potential solution from payload offering a slightly more secure way of handling this.

If however this isn't possible across domains thats no stress, the docs should probably just outline that accordingly along with the example code.

**Comment by prowsejeremy** [comment]
@jmikrut, appreciate your response, however, as per the wider conversation of this ticket / thread:

I am aware that passing through the token as part of the URL string is possible. It's obvious and easy enough to modify the returned preview url for the `preview button` that is rendered on a collection in the CMS. However this raises some security concerns having the token so visible in the request url.
Sure, using a request cookie is not a bulletproof solution, however it is at least another layer obfuscated from the URL itself. Adding the token to the request headers sent when requesting a preview would indeed be another viable solution, however this remains rather unclear as to how you would implement such a header. From what I can see (according to the docs and doing a little digging through the payload source) the only values made available upon configuring the preview button are the `doc` in question and the `GeneratePreviewURLOptions` which just pass back the current locale and token:

Payload src def for the preview url:

```
// .../payload/dist/config/types.d.ts (line 60)
export type GeneratePreviewURL = (doc: Record<string, unknown>, options: GeneratePreviewURLOptions) => Promise<null | string> | null | string;
```

And the Payload docs outline:
https://payloadcms.com/docs/configuration/collections#preview

I have configured JWT auth tokens (for preview data) in such a way in other CMS solutions and it does work adequately. There is just (again, from what I can see) no option to configure or pass through any custom header properties as you are suggesting in relation to the draft preview functionality built into Payload.


All of this to say, whatever the `Payload` way of handling this is, the main point I am raising is that the docs seem rather unclear as to how you should handle `draft preview` across differing domains. Some form of outline in the docs / example code would be very beneficial in helping people configure this "correctly".

Many thanks.

**Comment by jmikrut** [comment]
Ah @prowsejeremy good call. I forgot to mention that you can also override the entire preview button itself rather than just passing it a function to create the URL.

With that custom preview button, you can create your own click handler with your own fetch function to pass the token in a post request body or as a header.

https://payloadcms.com/docs/admin/components#collections

@kendelljoseph let’s update the docs to include an example of how to replace the entire preview button with your own custom button that passes the token securely outside of the URL in a safer manner.

@prowsejeremy thanks for following up here!

**Comment by prowsejeremy** [comment]
No problem @jmikrut love the work ya'll are doing on this platform, keep up the awesome job!
